File: shared/projectTree.ts
/**
 * Pure tree-building logic for the project sidebar.
 * Shared between the web UI and TUI — no React, no DOM.
 */

import type { SessionInfo } from "./types.js";
import { projectKey, projectDisplayName, shortPath } from "./format.js";

// ---- ProjectNode ----

export interface ProjectNode {
  name: string;
  key: string;
  sessionCount: number;
  hasOngoing: boolean;
}

export function buildProjectNodes(sessions: SessionInfo[]): ProjectNode[] {
  const map = new Map<string, ProjectNode>();
  for (const s of sessions) {
    const key = projectKey(s.path);
    const existing = map.get(key);
    if (existing) {
      existing.sessionCount++;
      if (s.is_ongoing) existing.hasOngoing = true;
    } else {
      map.set(key, {
        name: shortPath(s.cwd) || projectDisplayName(key),
        key,
        sessionCount: 1,
        hasOngoing: s.is_ongoing,
      });
    }
  }
  return [...map.values()].toSorted((a, b) => a.name.localeCompare(b.name));
}

// ---- TreeNode ----

export interface TreeNode {
  node: ProjectNode;
  children: TreeNode[];
}

export function treeNodeCmp(a: TreeNode, b: TreeNode): number {
  return a.node.name.localeCompare(b.node.name);
}

export function buildTree(nodes: ProjectNode[]): TreeNode[] {
  const sorted = [...nodes].toSorted((a, b) => a.key.length - b.key.length);
  const roots: TreeNode[] = [];
  const all: TreeNode[] = [];

  for (const node of sorted) {
    let parent: TreeNode | undefined;
    for (const candidate of all) {
      if (
        node.key.startsWith(candidate.node.key + "-") &&
        (!parent || candidate.node.key.length > parent.node.key.length)
      ) {
        parent = candidate;
      }
    }

    const tn: TreeNode = { node, children: [] };
    all.push(tn);

    if (parent) {
      // Split the suffix by "--" to find intermediate path segments.
      // e.g. parent="-Users-me-proj", child="-Users-me-proj-svc--claude-worktrees-foo"
      // suffix="-svc--claude-worktrees-foo" → parts=["-svc","claude-worktrees-foo"]
      // intermediates=["svc"] → virtual node inserted between parent and child.
      const suffix = node.key.slice(parent.node.key.length);
      const suffixParts = suffix.split("--");
      const intermediates = suffixParts
        .slice(0, -1)
        .map((s) => s.replace(/^-+/, ""))
        .filter((s) => s.length > 0);

      if (intermediates.length > 0) {
        let attachTo = parent;
        for (const seg of intermediates) {
          const virtualKey = `__virtual:${attachTo.node.key}:${seg}`;
          let vn = attachTo.children.find((c) => c.node.key === virtualKey);
          if (!vn) {
            vn = {
              node: { name: seg, key: virtualKey, sessionCount: 0, hasOngoing: false },
              children: [],
            };
            attachTo.children.push(vn);
            all.push(vn);
          }
          vn.node = {
            ...vn.node,
            sessionCount: vn.node.sessionCount + node.sessionCount,
            hasOngoing: vn.node.hasOngoing || node.hasOngoing,
          };
          attachTo = vn;
        }
        attachTo.children.push(tn);
      } else {
        const childLabel =
          node.key.slice(parent.node.key.length).replace(/^-+/, "") || projectDisplayName(node.key);
        tn.node = { ...node, name: node.name || childLabel };
        parent.children.push(tn);
      }
    } else {
      roots.push(tn);
    }
  }

  roots.sort(treeNodeCmp);
  for (const r of all) r.children.sort(treeNodeCmp);
  return roots;
}

// ---- Worktree helpers ----

export type WorktreeKind = "worktrees" | "claude-worktrees";

export function detectWorktreeKind(parentKey: string, childKey: string): WorktreeKind | null {
  // New style: key contains "--" — detect from the last "--"-separated segment.
  const parts = childKey.split("--");
  if (parts.length > 1) {
    const last = parts[parts.length - 1];
    if (last.startsWith("claude-worktrees")) return "claude-worktrees";
    if (last.startsWith("worktrees")) return "worktrees";
    return null;
  }
  // Old style: single-dash suffix relative to parent key.
  const rest = childKey.slice(parentKey.length);
  if (rest.startsWith("--claude-worktrees-")) return "claude-worktrees";
  if (rest.startsWith("-worktrees-")) return "worktrees";
  return null;
}

export function worktreeLeafName(parentKey: string, childKey: string, kind: WorktreeKind): string {
  // New style: extract leaf from last "--"-separated segment.
  const parts = childKey.split("--");
  if (parts.length > 1) {
    const last = parts[parts.length - 1];
    const prefix = kind === "claude-worktrees" ? "claude-worktrees-" : "worktrees-";
    return last.startsWith(prefix) ? last.slice(prefix.length) : projectDisplayName(childKey);
  }
  // Old style: slice past the worktree prefix from the parent key.
  const prefixLen =
    kind === "claude-worktrees" ? "--claude-worktrees-".length : "-worktrees-".length;
  return (
    childKey.slice(parentKey.length + prefixLen).replace(/^-+/, "") || projectDisplayName(childKey)
  );
}

// ---- FlatItem ----

export interface FlatItem {
  key: string | null;
  name: string;
  count: number;
  ongoing: boolean;
  depth: number;
  isGroup: boolean;
  hasChildren: boolean;
  isExpanded: boolean;
}

export function flattenTree(roots: TreeNode[], collapsedKeys?: ReadonlySet<string>): FlatItem[] {
  const items: FlatItem[] = [];

  function walk(nodes: TreeNode[], depth: number) {
    for (const tn of nodes) {
      const isCollapsed = collapsedKeys?.has(tn.node.key) ?? false;
      const hasChildren = tn.children.length > 0;

      items.push({
        key: tn.node.key,
        name: tn.node.name,
        count: tn.node.sessionCount,
        ongoing: tn.node.hasOngoing,
        depth,
        isGroup: false,
        hasChildren,
        isExpanded: !isCollapsed,
      });

      if (isCollapsed) continue;

      // Categorise children into worktree groups and regular
      const groups = new Map<WorktreeKind, TreeNode[]>();
      const regular: TreeNode[] = [];
      for (const child of tn.children) {
        const kind = detectWorktreeKind(tn.node.key, child.node.key);
        if (kind) {
          let list = groups.get(kind);
          if (!list) {
            list = [];
            groups.set(kind, list);
          }
          list.push(child);
        } else {
          regular.push(child);
        }
      }

      // Emit worktree group headers + their children
      for (const [kind, children] of groups) {
        const totalCount = children.reduce((s, c) => s + c.node.sessionCount, 0);
        const anyOngoing = children.some((c) => c.node.hasOngoing);
        const groupKey = `__group:${kind}:${tn.node.key}`;
        const groupIsCollapsed = collapsedKeys?.has(groupKey) ?? false;

        items.push({
          key: groupKey,
          name: kind,
          count: totalCount,
          ongoing: anyOngoing,
          depth: depth + 1,
          isGroup: true,
          hasChildren: children.length > 0,
          isExpanded: !groupIsCollapsed,
        });

        if (!groupIsCollapsed) {
          for (const child of children) {
            const childIsCollapsed = collapsedKeys?.has(child.node.key) ?? false;
            items.push({
              key: child.node.key,
              name: worktreeLeafName(tn.node.key, child.node.key, kind),
              count: child.node.sessionCount,
              ongoing: child.node.hasOngoing,
              depth: depth + 2,
              isGroup: false,
              hasChildren: child.children.length > 0,
              isExpanded: !childIsCollapsed,
            });
            if (!childIsCollapsed) walk(child.children, depth + 3);
          }
        }
      }

      walk(regular, depth + 1);
    }
  }

  walk(roots, 0);
  return items;
}

// ---- Convenience ----

/** Chains buildProjectNodes -> buildTree -> flattenTree and prepends "All Projects". */
export function buildFlatItems(
  sessions: SessionInfo[],
  collapsedKeys?: ReadonlySet<string>,
): FlatItem[] {
  const nodes = buildProjectNodes(sessions);
  const tree = buildTree(nodes);
  const flat = flattenTree(tree, collapsedKeys);
  return [
    {
      key: null,
      name: "All Projects",
      count: sessions.length,
      ongoing: false,
      depth: 0,
      isGroup: false,
      hasChildren: false,
      isExpanded: true,
    },
    ...flat,
  ];
}

File: tui-py/project_tree.py
"""Project tree building logic — Python port of shared/projectTree.ts."""

from __future__ import annotations

from dataclasses import dataclass, field

from data_types import FlatItem, SessionInfo
from format_utils import project_display_name, project_key, short_path

# ---------------------------------------------------------------------------
# ProjectNode
# ---------------------------------------------------------------------------


@dataclass
class ProjectNode:
    name: str
    key: str
    session_count: int
    has_ongoing: bool


def build_project_nodes(sessions: list[SessionInfo]) -> list[ProjectNode]:
    mapping: dict[str, ProjectNode] = {}
    for s in sessions:
        key = project_key(s.path)
        if key in mapping:
            mapping[key].session_count += 1
            if s.is_ongoing:
                mapping[key].has_ongoing = True
        else:
            name = short_path(s.cwd) or project_display_name(key)
            mapping[key] = ProjectNode(
                name=name,
                key=key,
                session_count=1,
                has_ongoing=s.is_ongoing,
            )
    return sorted(mapping.values(), key=lambda n: n.name)


# ---------------------------------------------------------------------------
# TreeNode
# ---------------------------------------------------------------------------


@dataclass
class TreeNode:
    node: ProjectNode
    children: list[TreeNode] = field(default_factory=list)


def build_tree(nodes: list[ProjectNode]) -> list[TreeNode]:
    sorted_nodes = sorted(nodes, key=lambda n: len(n.key))
    roots: list[TreeNode] = []
    all_nodes: list[TreeNode] = []

    for node in sorted_nodes:
        parent: TreeNode | None = None
        for candidate in all_nodes:
            if node.key.startswith(candidate.node.key + "-") and (
                parent is None or len(candidate.node.key) > len(parent.node.key)
            ):
                parent = candidate

        tn = TreeNode(node=node)
        all_nodes.append(tn)

        if parent is not None:
            suffix = node.key[len(parent.node.key) :]
            suffix_parts = suffix.split("--")
            intermediates = [p.lstrip("-") for p in suffix_parts[:-1] if p.lstrip("-")]

            if intermediates:
                attach_to = parent
                for seg in intermediates:
                    virtual_key = f"__virtual:{attach_to.node.key}:{seg}"
                    vn = next(
                        (c for c in attach_to.children if c.node.key == virtual_key),
                        None,
                    )
                    if vn is None:
                        vn = TreeNode(
                            node=ProjectNode(
                                name=seg,
                                key=virtual_key,
                                session_count=0,
                                has_ongoing=False,
                            )
                        )
                        attach_to.children.append(vn)
                        all_nodes.append(vn)
                    vn.node.session_count += node.session_count
                    vn.node.has_ongoing = vn.node.has_ongoing or node.has_ongoing
                    attach_to = vn
                attach_to.children.append(tn)
            else:
                child_label = node.key[len(parent.node.key) :].lstrip("-") or project_display_name(
                    node.key
                )
                tn.node = ProjectNode(
                    name=node.name or child_label,
                    key=node.key,
                    session_count=node.session_count,
                    has_ongoing=node.has_ongoing,
                )
                parent.children.append(tn)
        else:
            roots.append(tn)

    roots.sort(key=lambda t: t.node.name)
    for r in all_nodes:
        r.children.sort(key=lambda t: t.node.name)

    return roots


# ---------------------------------------------------------------------------
# Worktree helpers
# ---------------------------------------------------------------------------

WorktreeKind = str  # "worktrees" | "claude-worktrees"


def detect_worktree_kind(parent_key: str, child_key: str) -> WorktreeKind | None:
    parts = child_key.split("--")
    if len(parts) > 1:
        last = parts[-1]
        if last.startswith("claude-worktrees"):
            return "claude-worktrees"
        if last.startswith("worktrees"):
            return "worktrees"
        return None
    rest = child_key[len(parent_key) :]
    if rest.startswith("--claude-worktrees-"):
        return "claude-worktrees"
    if rest.startswith("-worktrees-"):
        return "worktrees"
    return None


def worktree_leaf_name(parent_key: str, child_key: str, kind: WorktreeKind) -> str:
    parts = child_key.split("--")
    if len(parts) > 1:
        last = parts[-1]
        prefix = "claude-worktrees-" if kind == "claude-worktrees" else "worktrees-"
        if last.startswith(prefix):
            return last[len(prefix) :]
        return project_display_name(child_key)
    prefix_len = len("--claude-worktrees-") if kind == "claude-worktrees" else len("-worktrees-")
    return child_key[len(parent_key) + prefix_len :].lstrip("-") or project_display_name(child_key)


# ---------------------------------------------------------------------------
# FlatItem flattening
# ---------------------------------------------------------------------------


def flatten_tree(roots: list[TreeNode], collapsed_keys: set[str] | None = None) -> list[FlatItem]:
    items: list[FlatItem] = []
    if collapsed_keys is None:
        collapsed_keys = set()

    def walk(nodes: list[TreeNode], depth: int) -> None:
        for tn in nodes:
            is_collapsed = tn.node.key in collapsed_keys
            has_children = len(tn.children) > 0

            items.append(
                FlatItem(
                    key=tn.node.key,
                    name=tn.node.name,
                    count=tn.node.session_count,
                    ongoing=tn.node.has_ongoing,
                    depth=depth,
                    is_group=False,
                    has_children=has_children,
                    is_expanded=not is_collapsed,
                )
            )

            if is_collapsed:
                continue

            # Categorise children
            groups: dict[str, list[TreeNode]] = {}
            regular: list[TreeNode] = []
            for child in tn.children:
                kind = detect_worktree_kind(tn.node.key, child.node.key)
                if kind:
                    groups.setdefault(kind, []).append(child)
                else:
                    regular.append(child)

            for kind, children in groups.items():
                total_count = sum(c.node.session_count for c in children)
                any_ongoing = any(c.node.has_ongoing for c in children)
                group_key = f"__group:{kind}:{tn.node.key}"
                group_collapsed = group_key in collapsed_keys

                items.append(
                    FlatItem(
                        key=group_key,
                        name=kind,
                        count=total_count,
                        ongoing=any_ongoing,
                        depth=depth + 1,
                        is_group=True,
                        has_children=len(children) > 0,
                        is_expanded=not group_collapsed,
                    )
                )

                if not group_collapsed:
                    for child in children:
                        child_collapsed = child.node.key in collapsed_keys
                        items.append(
                            FlatItem(
                                key=child.node.key,
                                name=worktree_leaf_name(tn.node.key, child.node.key, kind),
                                count=child.node.session_count,
                                ongoing=child.node.has_ongoing,
                                depth=depth + 2,
                                is_group=False,
                                has_children=len(child.children) > 0,
                                is_expanded=not child_collapsed,
                            )
                        )
                        if not child_collapsed:
                            walk(child.children, depth + 3)

            walk(regular, depth + 1)

    walk(roots, 0)
    return items


def build_flat_items(
    sessions: list[SessionInfo], collapsed_keys: set[str] | None = None
) -> list[FlatItem]:
    """Build the complete flat tree including 'All Projects' header."""
    nodes = build_project_nodes(sessions)
    tree = build_tree(nodes)
    flat = flatten_tree(tree, collapsed_keys)
    return [
        FlatItem(
            key=None,
            name="All Projects",
            count=len(sessions),
            ongoing=False,
            depth=0,
            is_group=False,
            has_children=False,
            is_expanded=True,
        ),
        *flat,
    ]
